### DISCUSSION TITLE
Access JS SDK `beforeSend` headers in API rules

### DISCUSSION THREAD

**Discussion by charbs-io** [pocketbase/pocketbase#1930] [Q&A] [answered]
Hi,

I want to use the js-sdk's `beforeSend` hook to append a value to the headers.

The usecase is limiting what data the user sees depending on which "workspace" they're currently in.

Is there any way headers can be accessed via the API rules in the Admin UI?

Thanks

**Comment by ganigeorgiev** [comment, marked as answer]
> Is there any way headers can be accessed via the API rules in the Admin UI?

No, this is not supported at the moment. We can add eventually support for this in the future, but for now as a workaround, you could pass the specific `workspace` value as query parameter and use the `@request.query.*` filter.

**Reply by charbs-io** [reply]
Thanks. But how can i pass the query value with the SDK? On the body?

**Reply by ganigeorgiev** [reply]
Hm, I forgot that the fetch query parameters are not part of the [options argument](https://developer.mozilla.org/en-US/docs/Web/API/fetch#options).

Indeed the query parameters at the moment cannot be modified via the `pb.beforeSend` SDK hook. The body parameters should work fine (keep in mind the body is serialized json or `FormData`).

Sometime later today I'll consider updating the hook to allow modifying the url string.

**Reply by charbs-io** [reply]
Okay thank you. The only thing thing with the body is you can't send it with a `get` request as far as i'm aware... Which is my main issue in this case (list and view apis needing to be limited to current workspace)

**Reply by charbs-io** [reply]
As a workaround i'm just going to use filters on the client side (i.e `filter: 'workspace_id = workspaceId'`) - but would be really nice to eventually also be able to lock it down from the server side.

**Reply by ganigeorgiev** [reply]
@cayoub88 You can add any query parameter when calling `getList`, `getOne`, etc., it is just not possible with the `pb.beforeSend` SDK hook (_I'll update it sometime later today_):

```js
await pb.collection("example").getList(1, 20, { "workspace_id": workspaceId });
```

**Reply by charbs-io** [reply]
> @cayoub88 You can add any query parameter when calling `getList`, `getOne`, etc., it is just not possible with the `pb.beforeSend` SDK hook (_I'll update it sometime later today_):
> 
> ```js
> await pb.collection("example").getList(1, 20, { "workspace_id": workspaceId });
> ```

Thanks. Yep i ended up using the filter to achieve a similar thing.

**Reply by ganigeorgiev** [reply]
@cayoub88 Just to let you know that in [v0.12.0 of the JS SDK](https://github.com/pocketbase/js-sdk/releases/tag/v0.12.0) it is now possible to modify the query parameters in the `pb.beforeSend` hook:

```js
pb.beforeSend = function (url, options) {
    const newUrl = new URL(url); // note that the it will throw if the client base url is not an absolute url
    newUrl.searchParams.set("workspace_id", "ABC123")

    return { url: newUrl.toString(), options }
}
```

_I'll consider also allowing access to the headers via the `@request.headers.*` filter in the next PocketBase v0.14.0 or v0.15.0 release._

**Reply by charbs-io** [reply]
@ganigeorgiev - amazing - really appreciate the speed of updating this! Thank you.
